File: exercises/bob/.meta/src/reference/java/Bob.java
/**
 * Bob is a lackadasical teenager.
 */
class Bob {
    String hey(String input) {
        input = normalize(input);
        if (isSilence(input)) {
            return "Fine. Be that way!";
        }
        if (isShout(input) && isQuestion(input)) {
            return "Calm down, I know what I'm doing!";
        }
        if (isShout(input)) {
            return "Whoa, chill out!";
        }
        if (isQuestion(input)) {
            return "Sure.";
        }
        return "Whatever.";
    }

    private static String normalize(String input) {
        return input.trim();
    }

    private static boolean isSilence(String input) {
        return input.equals("");
    }

    private static boolean isShout(String input) {
        final String upperCased = input.toUpperCase();
        final String lowerCased = input.toLowerCase();
        final boolean containsSomeLetters = !lowerCased.equals(upperCased);
        final boolean isAllUpperCase = upperCased.equals(input);
        return (containsSomeLetters && isAllUpperCase);
    }

    private static boolean isQuestion(String input) {
        return input.endsWith("?");
    }
}

File: go/bob/bob.go
// Package bob implements functionality of a robot that gives answers when prompted
package bob

import (
	"strings"
)

// Hey makes bob say hey on different ways based on a remark
func Hey(remark string) string {
	remark = strings.TrimSpace(remark)

	if remark == "" {
		return "Fine. Be that way!"
	}

	upper := strings.ToUpper(remark)
	lower := strings.ToLower(remark)

	if strings.HasSuffix(remark, "?") {
		if upper == remark && upper != lower {
			return "Calm down, I know what I'm doing!"
		}
		return "Sure."
	}
	if upper == remark && upper != lower {
		return "Whoa, chill out!"
	}

	return "Whatever."
}
